package org.purplebean.kmip.model.v2x1.structure.response.payload;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import lombok.Builder;
import lombok.Data;
import org.purplebean.kmip.api.EncodingType;
import org.purplebean.kmip.api.KmipContext;
import org.purplebean.kmip.api.KmipDataType;
import org.purplebean.kmip.api.KmipSpec;
import org.purplebean.kmip.api.KmipTag;
import org.purplebean.kmip.api.response.ResponsePayloadStructure;
import org.purplebean.kmip.model.core.enumeration.Operation;
import org.purplebean.kmip.model.core.enumeration.ValidityIndicator;
import org.purplebean.kmip.model.core.type.UniqueIdentifier;
import org.purplebean.kmip.model.v2x1.type.CorrelationValue;

@Data
@Builder(toBuilder = true)
public class MacVerifyOpResponsePayload implements ResponsePayloadStructure {

  private static final Operation.Value operation = Operation.Standard.MAC_VERIFY;
  private static final Set<KmipSpec> supportedVersions =
      Set.of(KmipSpec.UnknownVersion, KmipSpec.V2_1, KmipSpec.V3_0);

  static {
    for (KmipSpec spec : supportedVersions) {
      if (spec == KmipSpec.UnknownVersion || spec == KmipSpec.UnsupportedVersion) {
        continue;
      }
      KmipDataType.register(spec, kmipTag.getValue(), encodingType,
          MacVerifyOpResponsePayload.class);
      ResponsePayloadStructure.register(spec, operation, MacVerifyOpResponsePayload.class,
          MacVerifyOpResponsePayload::of);
    }
  }

  private final UniqueIdentifier uniqueIdentifier;
  private final ValidityIndicator validityIndicator;
  private final CorrelationValue correlationValue;

  @Builder
  private MacVerifyOpResponsePayload(
      UniqueIdentifier uniqueIdentifier,
      ValidityIndicator validityIndicator,
      CorrelationValue correlationValue
  ) {
    this.uniqueIdentifier = uniqueIdentifier;
    this.validityIndicator = validityIndicator;
    this.correlationValue = correlationValue;
    validate();
  }

  public static MacVerifyOpResponsePayload of(List<KmipDataType> values) {
    var builder = MacVerifyOpResponsePayload.builder();
    Map<KmipTag, List<KmipDataType>> map = values
        .stream()
        .collect(Collectors.groupingBy(KmipDataType::getKmipTag));
    if (map.containsKey(UniqueIdentifier.kmipTag)) {
      builder.uniqueIdentifier((UniqueIdentifier) map
          .get(UniqueIdentifier.kmipTag)
          .getFirst());
    }
    if (map.containsKey(ValidityIndicator.kmipTag)) {
      builder.validityIndicator((ValidityIndicator) map
          .get(ValidityIndicator.kmipTag)
          .getFirst());
    }
    if (map.containsKey(CorrelationValue.kmipTag)) {
      builder.correlationValue((CorrelationValue) map
          .get(CorrelationValue.kmipTag)
          .getFirst());
    }
    return builder.build();
  }

  private void validate() {
    if (!isSupported()) {
      throw new IllegalArgumentException(
          String.format("Unsupported object type for %s: %s", KmipContext.getSpec(), getKmipTag()));
    }
  }

  @Override
  public KmipTag getKmipTag() {
    return kmipTag;
  }

  @Override
  public EncodingType getEncodingType() {
    return encodingType;
  }

  @Override
  public boolean isSupported() {
    KmipSpec spec = KmipContext.getSpec();
    return supportedVersions.contains(spec) && Stream
        .of(getValue())
        .allMatch(KmipDataType::isSupported);
  }

  @Override
  public KmipDataType[] getValue() {
    return Stream
        .of(uniqueIdentifier, validityIndicator, correlationValue)
        .filter(Objects::nonNull)
        .map(kmipDataType -> kmipDataType)
        .toArray(KmipDataType[]::new);
  }

  @Override
  public Operation getCorrespondingOperation() {
    return operation.inst();
  }
}
